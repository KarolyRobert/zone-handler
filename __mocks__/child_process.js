'use strict';

const { EventEmitter, Readable, Writable } = require('stream');
const child_process = jest.createMockFromModule('child_process');

class WriteMemory extends Writable {
    constructor() {
      super();
      this.buffer = "";
    }
  
    _write(chunk, _, next) {
      this.buffer += chunk;
      next();
    }
  
    reset() {
      this.buffer = "";
    }
  }

child_process._current;

child_process.spawn = jest.fn(() => {
    const cprocess = new EventEmitter();
    cprocess.stdin = new WriteMemory();
    cprocess.stdout = new Readable();
    cprocess.stderr = new Readable();



    cprocess.stdout._read = cprocess.stderr._read = (size) => {};
   
    child_process._current = cprocess;
    return cprocess;
});

module.exports = child_process;