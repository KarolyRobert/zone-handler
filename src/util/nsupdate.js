import child_process from "child_process";

export default function nsupdate(commands){
    return new Promise((resolve,reject) => {
        let output = '';
        const nsupdateProcess = child_process.spawn('nsupdate',['-v']);
 
        nsupdateProcess.stdout.on('data', data => {
            output += data.toString();
        });
        nsupdateProcess.stderr.on('data', err => {
            reject(err.toString());
        });
        nsupdateProcess.on('close', code => {
        
            if(code === 0){
                resolve(output);
            }else{
                reject(`nsupdate exit with ${code}`);
            }

        });
        nsupdateProcess.on('error',err => {
            reject(err.message);
        });
        process.nextTick(() => {
            nsupdateProcess.stdin.end(commands);
        });
    });
}