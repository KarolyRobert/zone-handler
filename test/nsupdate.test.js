import child_process from 'child_process';
import nsupdate from '../src/util/nsupdate';

jest.mock('child_process');

const updateCommand =  'zone hobbyfork.com\n'+
'key hmac-sha256:update.key asdfksf8s6s875g765\n'+
'update delete www.hobbyfork.com. 30 A 37.234.91.82\n'+
'update add home.hobbyfork.com. 21600 A 192.168.1.6\n'+
'send\n'

describe('nsupdate',() => {

    test('nsupdate call child_process.spawn with given parameters.',() => {
        expect.assertions(3);
        nsupdate(updateCommand);
        expect(child_process.spawn.mock.calls.length).toBe(1);
        expect(child_process.spawn.mock.calls[0][0]).toBe('nsupdate');
        expect(child_process.spawn.mock.calls[0][1]).toEqual(['-v']);
    });

    test('nsupdate reject if bindutils/nsupdate not in running environment.',() => {
        expect.assertions(1);
        nsupdate(updateCommand).catch(err => {
            expect(err).toBe('nsupdate error');
        });
        child_process._current.emit('error',new Error('nsupdate error'));
    });

    test('nsupdate reject on runtime error',done => {
        expect.assertions(1);
        let error;
            nsupdate(updateCommand).catch(err => {
                error = err;
                //expect(err).toBe('dig runtime error');
            });
        child_process._current.stderr.push('nsupdate runtime error');
        setTimeout(() => {
            expect(error).toBe('nsupdate runtime error');
            done();
        },100);
    });

    test('nsupdate reject on close not null exit code',() => {
        expect.assertions(1);
        nsupdate(updateCommand).catch(err => {
            expect(err).toBe('nsupdate exit with -1');
        });
        child_process._current.emit('close',-1);
    });

    test('nsupdate resolve on close code 0 with proper result',done => {
        expect.assertions(2);
        let result;
        nsupdate(updateCommand).then(res => {
            result = res;
        });
        
        process.nextTick(() => child_process._current.emit('close',0));
        setTimeout(() => {
            expect(result).toBe('');
            expect(child_process._current.stdin.buffer).toMatchSnapshot();
            done();
        },100);
    });

});