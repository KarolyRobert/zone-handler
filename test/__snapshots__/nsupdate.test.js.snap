// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nsupdate nsupdate resolve on close code 0 with proper result 1`] = `
"zone hobbyfork.com
key hmac-sha256:update.key asdfksf8s6s875g765
update delete www.hobbyfork.com. 30 A 37.234.91.82
update add home.hobbyfork.com. 21600 A 192.168.1.6
send
"
`;
